packages/frontend/src/model/theory_selector.tsx

```diff
@@ -48,22 +47,69 @@ export function TheorySelector(props: TheorySelectorProps) {
     const groupedTheories = createMemo(() =>
         Array.from(props.theories.groupedMetadata().entries()),
     );
-
     return (
         <div class="theory-selector">
             <For each={groupedTheories()}>
                 {([group, theories]) => (
                     <div class="group">
                         <div class="group-name">{group}</div>
-                        <For each={theories}>
+                        <For
+                            each={theories.filter((t) => {
+                                if (props.formalCells.length === 0) {
+                                    return true;
+                                }
+                                const fwd = props.theories
+                                    .get(props.theory.id)
+                                    .inclusions.get(t.id);
+                                if (!(fwd === undefined)) {
+                                    return true;
+                                }
+
+                                const bkwd = props.theories
+                                    .get(t.id)
+                                    .inclusions.get(props.theory.id);
+
+                                if (!(bkwd === undefined)) {
+                                    return props.formalCells.every((fc) => {
+                                        return (
+                                            Array.from(bkwd.obnames.values()).includes(fc) ||
+                                            Array.from(bkwd.mornames.values()).includes(fc)
+                                        );
+                                    });
+                                }
+                                return false;
+                            })}
+                        >
                             {(meta) => (
                                 <div class="theory">
                                     <input
                                         type="radio"
                                         name="theory"
                                         id={meta.id}
                                         value={meta.id}
-                                        onchange={(evt) => props.setTheory(evt.target.value)}
+                                        onchange={(evt) => {
```

Comment:
Ditto about big event handlers: better to make the separate helper functions/closures rather than inline JSX.